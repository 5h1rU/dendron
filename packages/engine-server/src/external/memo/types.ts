import { URI } from "vscode-uri";

export type WorkspaceCache = {
  imageUris: URI[];
  markdownUris: URI[];
  otherUris: URI[];
  allUris: URI[];
  danglingRefs: string[];
  danglingRefsByFsPath: { [key: string]: string[] };
};

export type RefT = {
  label: string;
  ref: string;
};
