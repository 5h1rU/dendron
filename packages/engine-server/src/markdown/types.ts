import { DNoteRefLink, NotePropsV2 } from "@dendronhq/common-all";
import MDAST from "mdast";

export type DendronASTNode = MDAST.Parent & {
  notes?: NotePropsV2[];
  children: MDAST.Parent["children"] | DendronASTNode[];
};

export enum DendronASTTypes {
  WIKI_LINK = "wikiLink",
  REF_LINK = "refLink",
  PARAGRAPH = "paragraph",
}

export enum DendronASTDest {
  // MD_ENHANCED_PREVIEW = "MD_ENHANCED_PREVIEW",
  MD_REGULAR = "MD_REGULAR",
  HTML = "HTML",
}

export type DendronASTData = {
  dest: DendronASTDest;
};

export type WikiLinkNoteV4 = DendronASTNode & {
  type: DendronASTTypes.WIKI_LINK;
  value: string;
  data: WikiLinkDataV4;
};

export type WikiLinkDataV4 = {
  alias: string;
  anchorHeader?: string;
  prefix?: string;
};

export type NoteRefNoteV4 = DendronASTNode & {
  type: DendronASTTypes.REF_LINK;
  value: string;
  data: NoteRefDataV4;
};

export type NoteRefDataV4 = {
  link: DNoteRefLink;
};
