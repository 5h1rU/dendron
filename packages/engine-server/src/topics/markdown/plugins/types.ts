import { DNoteLoc, DNoteRefLink } from "@dendronhq/common-all";
import { DendronASTNode, DendronASTTypes } from "../../../markdown/types";

// --- General

// @deprecate
export type LinkOpts = {
  toMdEnhancedPreview?: boolean;
};

// --- WIKI Link
export type RefLinkNote = DendronASTNode & {
  type: DendronASTTypes.REF_LINK;
  data: RefLinkData;
  value: string;
};

export type RefLinkData = {
  link: DNoteRefLink;
  // --- Old
  alias: string;
  permalink: string;
  exists: boolean;
  hName: string;
  hProperties: any;
  hChildren: any[];
  toMd?: boolean;
  prefix?: string;
  useId: boolean;
};

export type WikiLinkNote = DendronASTNode & {
  type: DendronASTTypes.WIKI_LINK;
  data: WikiLinkData;
  value: string;
};

export type WikiLinkData = {
  alias: string;
  permalink: string;
  exists: boolean;
  hName: string;
  hProperties: {
    className: string;
    href: string;
  };
  hChildren: any[];
  anchorHeader?: string;

  // dendron specific
  toMd?: boolean;
  toHTML?: boolean;
  /**
   * {ROOT_URL}/{prefix?}/{destination}
   */
  prefix?: string;
  useId: boolean;
  note?: { id: string };
  replace?: DNoteLoc;
  forNoteRefInPreview?: boolean;
  forNoteRefInSite?: boolean;
} & LinkOpts;

export type WikiLinkProps = {
  alias: string;
  value: string;
  anchorHeader?: string;
  filters?: LinkFilter[];
};

export type LinkFilter = {
  name: string;
};
