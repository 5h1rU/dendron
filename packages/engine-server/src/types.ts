import { DEngineClientV2, DNoteLoc } from "@dendronhq/common-all";

export type ReplaceLinkOpts = { from: DNoteLoc; to: DNoteLoc };
export { DEngineClientV2 };
