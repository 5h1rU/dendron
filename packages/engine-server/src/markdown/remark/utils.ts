import {
  CONSTANTS,
  DendronError,
  getSlugger,
  NoteChangeEntry,
  NoteProps,
  NoteUtils,
} from "@dendronhq/common-all";
import { normalizev2 } from "@dendronhq/engine-server";
import _ from "lodash";
import { Heading, Root } from "mdast";
import { Processor } from "unified";
import { Node } from "unist";
import { selectAll } from "unist-util-select";
import { VFile } from "vfile";
import { WikiLinkProps } from "../../topics/markdown";
import {
  DendronASTRoot,
  DendronASTTypes,
  NoteRefNoteV4_LEGACY,
} from "../types";

export const ALIAS_DIVIDER = "|";

export function addError(proc: Processor, err: DendronError) {
  const errors = proc.data("errors") as DendronError[];
  errors.push(err);
  proc().data("errors", errors);
}

export function getNoteOrError(
  notes: NoteProps[],
  hint: any
): { error: DendronError | undefined; note: undefined | NoteProps } {
  let error: DendronError | undefined;
  let note: NoteProps | undefined;
  if (_.isUndefined(notes)) {
    error = new DendronError({ msg: `no note found. ${hint}` });
    return { error, note };
  }
  if (notes.length > 1) {
    error = new DendronError({ msg: `multiple notes found for link: ${hint}` });
    return { error, note };
  }
  if (notes.length < 1) {
    error = new DendronError({
      msg: `no notes found for link: ${JSON.stringify(hint)}`,
    });
    return { error, note };
  }
  note = notes[0];
  return { error, note };
}

export class LinkUtils {
  static isAlias(link: string) {
    return link.indexOf("|") !== -1;
  }

  static hasFilter(link: string) {
    return link.indexOf(">") !== -1;
  }

  static parseAliasLink(link: string) {
    const [alias, value] = link.split("|").map(_.trim);
    return { alias, value: NoteUtils.normalizeFname(value) };
  }

  static parseDendronURI(linkString: string) {
    if (linkString.startsWith(CONSTANTS.DENDRON_DELIMETER)) {
      const [vaultName, link] = linkString
        .split(CONSTANTS.DENDRON_DELIMETER)[1]
        .split("/");
      return {
        vaultName,
        link,
      };
    }
    return {
      link: linkString,
    };
  }

  static parseLinkV2(linkString: string) {
    const LINK_NAME = "[^#\\|>]+";
    const re = new RegExp(
      "" +
        // alias?
        `(` +
        `(?<alias>${LINK_NAME}(?=\\|))` +
        "\\|" +
        ")?" +
        // name
        `(?<value>${LINK_NAME})` +
        // anchor?
        `(#(?<anchor>${LINK_NAME}))?` +
        // filters?
        `(>(?<filtersRaw>.*))?`,
      "i"
    );
    const out = linkString.match(re);
    if (out) {
      let { alias, value, anchor } = out.groups as any;
      let vaultName: string | undefined;
      ({ vaultName, link: value } = this.parseDendronURI(value));
      if (!alias) {
        alias = value;
      }
      alias = _.trim(alias);
      value = _.trim(value);
      return { alias, value, anchorHeader: anchor, vaultName };
    } else {
      return null;
    }
  }

  static parseLink(linkMatch: string) {
    linkMatch = NoteUtils.normalizeFname(linkMatch);
    let out: WikiLinkProps = {
      value: linkMatch,
      alias: linkMatch,
    };
    if (LinkUtils.isAlias(linkMatch)) {
      out = LinkUtils.parseAliasLink(linkMatch);
    }
    if (out.value.indexOf("#") !== -1) {
      const [value, anchorHeader] = out.value.split("#").map(_.trim);
      out.value = value;
      out.anchorHeader = anchorHeader;
      // if we didn't have an alias, links with a # anchor shouldn't have # portion be in the title
      if (!LinkUtils.isAlias(linkMatch)) {
        out.alias = value;
      }
    }
    return out;
  }
}

export class RemarkUtils {
  static oldNoteRef2NewNoteRef(note: NoteProps, changes: NoteChangeEntry[]) {
    return function (this: Processor) {
      return (tree: Node, _vfile: VFile) => {
        let root = tree as DendronASTRoot;
        //@ts-ignore
        let notesRefLegacy: NoteRefNoteV4_LEGACY[] = selectAll("refLink", root);
        const slugger = getSlugger();
        notesRefLegacy.map((noteRefLegacy) => {
          // @ts-ignore;
          noteRefLegacy.type = DendronASTTypes.REF_LINK_V2;
          const { anchorStart, anchorEnd } = noteRefLegacy.data.link.data;
          if (anchorStart) {
            noteRefLegacy.data.link.data.anchorStart = normalizev2(
              anchorStart,
              slugger
            );
          }
          if (anchorEnd) {
            noteRefLegacy.data.link.data.anchorEnd = normalizev2(
              anchorEnd,
              slugger
            );
          }
        });
        changes.push({
          note,
          status: "update",
        });
        // if (out2.length > 0) {
        //   // const noteRefLegacy = root.children.splice(idx, 1)[0] as unknown as  NoteRefNoteV4_LEGACY;
        //   const noteRefLegacy = out2[0];
        //   // root.children[idx] as unknown as NoteRefNoteV4_LEGACY;
        //   // const noteRefNew: NoteRefNoteV4= {
        //   //   type: DendronASTTypes.REF_LINK_V2,
        //   //   data: {
        //   //     link: noteRefLegacy.data.link,
        //   //   },
        //   //   value: noteRefLegacy.value
        //   // }
        //   // @ts-ignore;
        //   noteRefLegacy.type = DendronASTTypes.REF_LINK_V2;
        //   // // @ts-ignore
        //   // root.children.splice(idx, 1, noteRefNew);
        //   changes.push({
        //     note,
        //     status: "update",
        //   });
        // }
      };
    };
  }

  static h1ToTitle(note: NoteProps, changes: NoteChangeEntry[]) {
    return function (this: Processor) {
      return (tree: Node, _vfile: VFile) => {
        let root = tree as Root;
        const idx = _.findIndex(
          root.children,
          (ent) => ent.type === "heading" && ent.depth === 1
        );
        if (idx >= 0) {
          const head = root.children.splice(idx, 1)[0] as Heading;
          if (head.children.length === 1 && head.children[0].type === "text") {
            note.title = head.children[0].value;
          }
          changes.push({
            note,
            status: "update",
          });
        }
      };
    };
  }

  static h1ToH2(note: NoteProps, changes: NoteChangeEntry[]) {
    return function (this: Processor) {
      return (tree: Node, _vfile: VFile) => {
        let root = tree as Root;
        const idx = _.findIndex(
          root.children,
          (ent) => ent.type === "heading" && ent.depth === 1
        );
        if (idx >= 0) {
          const head = root.children[idx] as Heading;
          head.depth = 2;
          changes.push({
            note,
            status: "update",
          });
        }
      };
    };
  }
}
