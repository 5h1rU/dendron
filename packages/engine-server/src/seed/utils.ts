import { DendronError, ERROR_STATUS, SeedConfig } from "@dendronhq/common-all";
import { simpleGit } from "@dendronhq/common-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";
import { WorkspaceService } from "../workspace";
import { DEFAULT_SEED_PUBLISHER } from "./constants";

export class SeedUtils {
  static genDefaultConfig(opts: {
    seed?: SeedConfig;
    name: string;
  }): SeedConfig {
    return _.defaults(opts.seed || {}, {
      name: opts.name,
      publisher: DEFAULT_SEED_PUBLISHER,
      description: "a seed waiting to sprout",
      license: "CC BY 4.0",
      repository: {
        type: "git" as const,
        url: "",
      },
      root: "vault",
    });
  }

  static async clone({
    wsRoot,
    config,
  }: {
    config: SeedConfig;
    wsRoot: string;
  }) {
    const id = this.getSeedId(config);
    const spath = this.seed2Path({ wsRoot, id });
    fs.ensureDirSync(path.dirname(spath));
    const git = simpleGit({ baseDir: wsRoot });
    await git.clone(config.repository.url, spath);
    return spath;
  }

  static getSeedId({ publisher, name }: { publisher: string; name: string }) {
    return `${publisher}.${name}`;
  }

  static seed2Path({ wsRoot, id }: { wsRoot: string; id: string }) {
    return path.join(wsRoot, "seeds", id);
  }

  static validateWorkspaceSeedConversion({ wsRoot }: { wsRoot: string }) {
    const ws = new WorkspaceService({ wsRoot });
    if (ws.config.vaults.length !== 1) {
      return {
        error: DendronError.createFromStatus({
          status: ERROR_STATUS.INVALID_STATE,
          message: "workspace must have exactly one vault",
        }),
      };
    }
    if (!_.isEmpty(ws.config.workspaces)) {
      return {
        error: DendronError.createFromStatus({
          status: ERROR_STATUS.INVALID_STATE,
          message: "workspace vaults not supported",
        }),
      };
    }
    return {
      error: undefined,
    };
  }
}
