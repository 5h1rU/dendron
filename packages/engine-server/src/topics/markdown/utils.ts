/* eslint-disable camelcase */
import _markdownit from "markdown-it";
// @ts-ignore
import markdownItAST from "markdown-it-ast";
import markdownItRegex from "markdown-it-regex";
import Token from "markdown-it/lib/token";
import { MDRenderer } from "./renderer";

enum NodeType {
  code_inline = "code_inline",
  code_block = "code_block",
  fence = "fence",
  image = "image",
  hardbreak = "hardbreak",
  softbreak = "softbreak",
  text = "text",
  html_block = "html_block",
  html_inline = "html_inline",
}

export type MDNode = {
  nodeType: NodeType;
  openNode: Token;
  closeNode: Token;
  children: Token[];
};

export function md() {
  return _markdownit();
}

function token2MD(token: Token) {
  switch (token.type) {
    case NodeType.text: {
      break;
    }
    default: {
      throw Error(`unhandled node type: ${token.type}`);
    }
  }
}

export function tokens2MD(tokens: Token[]): string {
  return new MDRenderer().renderInline(tokens, {}, {});
  //   return tokens
  //     .map((n) => {
  //       return token2MD(n);
  //     })
  //     .join("");
  //   const markdownIt = _markdownit();
  //   const renderer = new MDRenderer();
  //   const
  // @ts-ignore
  //   markdownIt.renderer = renderer;
  //   const options = {};
  //   const env = {};
  //     return renderer.render(nodes, {}, {});
  //   return markdownIt.render(nodes as string);
}

export function md2Tokens(txt: string): Token[] {
  const markdownIt = _markdownit();
  const tokens: Token[] = markdownIt
    .use(markdownItRegex, {
      name: "ref-document",
      regex: /\[\[([^\[\]]+?)\]\]/,
      // replace: (rawRef: string) => {
      //   const { ref, label } = parseRef(rawRef);
      //   // const fsPath = findUriByRef(getWorkspaceCache().allUris, ref)?.fsPath;
      //   // if (!fsPath) {
      //   //   return getInvalidRefAnchor(label || ref);
      //   // }
      //   return getRefAnchor(ref, label || ref);
      // },
    })
    .parse(txt, {});
  return tokens;
}
