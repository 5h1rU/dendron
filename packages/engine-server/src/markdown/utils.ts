import { DendronError, DEngineClientV2, DVault } from "@dendronhq/common-all";
import _ from "lodash";
import { Heading } from "mdast";
import { paragraph, root, text } from "mdast-builder";
// @ts-ignore
import katex from "rehype-katex";
import raw from "rehype-raw";
import rehypeStringify from "rehype-stringify";
import remark from "remark";
import frontmatterPlugin from "remark-frontmatter";
import math from "remark-math";
import remarkParse from "remark-parse";
import remark2rehype from "remark-rehype";
import { default as unified, default as Unified, Processor } from "unified";
import { Node, Parent } from "unist";
import { dendronPub, DendronPubOpts } from "./remark/dendronPub";
import { noteRefs, NoteRefsOpts } from "./remark/noteRefs";
import { wikiLinks, WikiLinksOpts } from "./remark/wikiLinks";
import { DendronASTData, DendronASTDest } from "./types";
// @ts-ignore
import highlight from "remark-highlight.js";

const toString = require("mdast-util-to-string");

type ProcOpts = {
  engine: DEngineClientV2;
};

export class MDUtilsV4 {
  static findIndex(array: Node[], fn: any) {
    for (var i = 0; i < array.length; i++) {
      if (fn(array[i], i)) {
        return i;
      }
    }
    return -1;
  }

  static genMDMsg(msg: string): Parent {
    return root(paragraph(text(msg)));
  }

  static getDendronData(proc: Processor) {
    return proc.data("dendron") as DendronASTData;
  }

  static setDendronData(proc: Processor, data: DendronASTData) {
    const _data = proc.data("dendron") as DendronASTData;
    return proc.data("dendron", { ..._data, ...data });
  }

  static getEngineFromProc(proc: Unified.Processor) {
    const engine = proc.data("engine") as DEngineClientV2;
    let error: DendronError | undefined;
    if (_.isUndefined(engine) || _.isNull(engine)) {
      error = new DendronError({ msg: "engine not defined" });
    }
    return {
      error,
      engine,
    };
  }

  static getNoteRefLvl(proc: Unified.Processor): number {
    return (proc.data("noteRefLvl") as number) || 0;
  }

  static setNoteRefLvl(proc: Unified.Processor, lvl: number) {
    return proc.data("noteRefLvl", lvl);
  }

  static isHeading(node: Node, text: string, depth?: number) {
    if (node.type !== "heading") {
      return false;
    }

    // wildcard is always true
    if (text === "*") {
      return true;
    }
    if (text) {
      var headingText = toString(node);
      return text.trim().toLowerCase() === headingText.trim().toLowerCase();
    }

    if (depth) {
      return (node as Heading).depth <= depth;
    }

    return true;
  }
  static proc(opts: ProcOpts) {
    const { engine } = opts;
    const errors: DendronError[] = [];
    let _proc = remark()
      .use(remarkParse, { gfm: true })
      .data("errors", errors)
      .data("engine", engine)
      .use(frontmatterPlugin, ["yaml"])
      .use({ settings: { listItemIndent: "1", fences: true, bullet: "-" } });
    return _proc;
  }

  static procFull(
    opts: ProcOpts & {
      dest: DendronASTDest;
      vault?: DVault;
      wikiLinksOpts?: WikiLinksOpts;
      noteRefOpts?: NoteRefsOpts;
      publishOpts?: DendronPubOpts;
      mathOpts?: {
        katex?: boolean;
      };
    }
  ) {
    const { dest, vault } = opts;
    let proc = this.proc(opts)
      .data("dendron", { dest, vault } as DendronASTData)
      .use(wikiLinks, opts.wikiLinksOpts)
      .use(noteRefs, { ...opts.noteRefOpts, wikiLinkOpts: opts.wikiLinksOpts });
    if (opts.mathOpts?.katex) {
      proc = proc.use(math);
    }
    // if (dest === DendronASTDest.HTML) {
    proc = proc.use(dendronPub, {
      ...opts.publishOpts,
      wikiLinkOpts: opts.wikiLinksOpts,
    });
    proc = proc.data("procFull", proc().freeze());
    return proc;
    //return proc;
  }

  static procRehype(opts: {
    proc?: Processor;
    mdPlugins?: Processor[];
    mathjax?: boolean;
  }) {
    const { proc, mdPlugins } = _.defaults(opts, { mdPlugins: [] });
    let _proc = proc || unified().use(remarkParse, { gfm: true });
    _proc = _proc.use(highlight);
    mdPlugins.forEach((p) => {
      _proc = _proc.use(p);
    });
    _proc = _proc.use(remark2rehype, { allowDangerousHtml: true }).use(raw);
    if (opts.mathjax) {
      _proc = _proc.use(katex);
    }
    return _proc.use(rehypeStringify);
  }
}
