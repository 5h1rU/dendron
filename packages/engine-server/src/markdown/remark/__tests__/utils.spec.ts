import { NotePropsV2 } from "@dendronhq/common-all";
import {
  AssertUtils,
  ENGINE_HOOKS,
  NoteTestUtilsV4,
  runEngineTestV4,
  TestPresetEntryV4,
} from "@dendronhq/common-test-utils";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";
import { DendronASTDest } from "../../types";
import { MDUtilsV4 } from "../../utils";
import { createEngine, createProcTests } from "./utils";

const IMAGE_LINK = `![alt-text](image-url.jpg)`;

const WITH_TITLE = createProcTests({
  name: "WITH_TITLE",
  setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
    const proc = await MDUtilsV4.procFull({
      engine,
      dest: extra.dest,
      fname: "foo",
      vault: vaults[0],
      publishOpts: {
        insertTitle: true,
      },
    });
    const noteRaw = fs.readFileSync(
      path.join(wsRoot, vaults[0].fsPath, "foo.md"),
      { encoding: "utf8" }
    );
    const resp = proc.processSync(noteRaw);
    const respParse = proc.parse(noteRaw);
    expect(resp).toMatchSnapshot();
    expect(respParse).toMatchSnapshot("respParse");
    return { resp, proc };
  },
  verifyFuncDict: {
    [DendronASTDest.MD_REGULAR]: async ({ extra }) => {
      const { resp } = extra;
      expect(
        await AssertUtils.assertInString({
          body: resp.contents,
          match: ["# Foo", "foo body"],
        })
      ).toBeTruthy();
    },
  },
  preSetupHook: ENGINE_HOOKS.setupBasic,
});

const WITH_VARIABLE = createProcTests({
  name: "WITH_VARIABLE",
  setupFunc: async ({ engine, vaults, extra, wsRoot }) => {
    let proc = await MDUtilsV4.procFull({
      engine,
      dest: extra.dest,
      fname: "foo",
      vault: vaults[0],
      publishOpts: {
        insertTitle: true,
      },
    });
    const noteRaw = fs.readFileSync(
      path.join(wsRoot, vaults[0].fsPath, "foo.md"),
      { encoding: "utf8" }
    );
    const resp = proc.processSync(noteRaw);
    const respParse = proc.parse(noteRaw);
    const respRehype = MDUtilsV4.procRehype({ proc: proc() }).processSync(
      noteRaw
    );
    expect(resp).toMatchSnapshot();
    expect(respParse).toMatchSnapshot("respParse");
    expect(respRehype).toMatchSnapshot("respRehype");
    return { resp, proc, respRehype };
  },
  verifyFuncDict: {
    [DendronASTDest.MD_REGULAR]: async ({ extra }) => {
      const { resp, respRehype } = extra;
      Promise.all(
        [resp, respRehype].map(async (ent) => {
          expect(
            await AssertUtils.assertInString({
              body: ent.contents,
              match: ["Title: Foo", "Bond: 42"],
            })
          ).toBeTruthy();
        })
      );
    },
  },
  preSetupHook: async (opts) => {
    await ENGINE_HOOKS.setupBasic(opts);
    await NoteTestUtilsV4.modifyNoteByPath(
      { wsRoot: opts.wsRoot, vault: opts.vaults[0], fname: "foo" },
      (note: NotePropsV2) => {
        note.custom = { bond: 42 };
        note.body = `Title: {{fm.title}}. Bond: {{fm.bond}}`;
        return note;
      }
    );
  },
});

const WITH_ASSET_PREFIX_UNDEFINED = createProcTests({
  name: "asset_prefix undefined",
  setupFunc: async ({ engine, vaults, extra }) => {
    const proc = await MDUtilsV4.procFull({
      engine,
      dest: extra.dest,
      vault: vaults[0],
      publishOpts: {
        assetsPrefix: undefined,
      },
    });
    const resp = proc.processSync(IMAGE_LINK);
    return { resp, proc };
  },
  verifyFuncDict: {
    [DendronASTDest.MD_REGULAR]: async ({ extra }) => {
      const { resp } = extra;
      expect(resp).toMatchSnapshot();
      return [
        {
          actual: _.trim(resp.toString()),
          expected: "![alt-text](image-url.jpg)",
        },
      ];
    },
    [DendronASTDest.HTML]: async ({ extra }) => {
      const { resp } = extra;
      expect(resp).toMatchSnapshot();
      return [
        {
          actual: _.trim(resp.toString()),
          expected: "![alt-text](image-url.jpg)",
        },
      ];
    },
    [DendronASTDest.MD_ENHANCED_PREVIEW]: async ({ extra }) => {
      const { resp } = extra;
      return [
        {
          actual: _.trim(resp.toString()),
          expected: "![alt-text](image-url.jpg)",
        },
      ];
    },
  },
  preSetupHook: ENGINE_HOOKS.setupBasic,
});

const WITH_ASSET_PREFIX = createProcTests({
  name: "asset_prefix",
  setupFunc: async ({ engine, vaults, extra }) => {
    const proc = await MDUtilsV4.procFull({
      engine,
      dest: extra.dest,
      vault: vaults[0],
      publishOpts: {
        assetsPrefix: "bond/",
      },
    });
    const resp = proc.processSync(IMAGE_LINK);
    return { resp, proc };
  },
  verifyFuncDict: {
    [DendronASTDest.MD_REGULAR]: async ({ extra }) => {
      const { resp } = extra;
      expect(resp).toMatchSnapshot();
      return [
        {
          actual: _.trim(resp.toString()),
          expected: "![alt-text](image-url.jpg)",
        },
      ];
    },
    [DendronASTDest.HTML]: async ({ extra }) => {
      const { resp } = extra;
      expect(resp).toMatchSnapshot();
      return [
        {
          actual: _.trim(resp.toString()),
          expected: "![alt-text](bond/image-url.jpg)",
        },
      ];
    },
    [DendronASTDest.MD_ENHANCED_PREVIEW]: async ({ extra }) => {
      const { resp } = extra;
      return [
        {
          actual: _.trim(resp.toString()),
          expected: "![alt-text](image-url.jpg)",
        },
      ];
    },
  },
  preSetupHook: ENGINE_HOOKS.setupBasic,
});

const NOTE_REF_BASIC_WITH_REHYPE = createProcTests({
  name: "NOTE_REF_WITH_REHYPE",
  setupFunc: async ({ engine, vaults, extra }) => {
    const proc = await MDUtilsV4.procFull({
      engine,
      wikiLinksOpts: { useId: true },
      dest: extra.dest,
      vault: vaults[0],
    });
    const txt = `((ref: [[foo.md]]))`;
    if (extra.dest === DendronASTDest.HTML) {
      const procRehype = MDUtilsV4.procRehype({ proc });
      const resp = await procRehype.process(txt);
      return { resp, proc };
    } else {
      const resp = await proc.process(txt);
      return { resp, proc };
    }
  },
  verifyFuncDict: {
    [DendronASTDest.MD_REGULAR]: async () => {},
    [DendronASTDest.HTML]: async ({ extra }) => {
      const { resp } = extra;
      expect(resp).toMatchSnapshot("respRehype");
      expect(
        await AssertUtils.assertInString({
          body: resp.contents,
          match: [
            // link by id
            `<a href=\"foo-id.html\"`,
            // html quoted
            `<p><a href=\"bar.html\">bar</a></p>`,
          ],
        })
      ).toBeTruthy();
    },
    [DendronASTDest.MD_ENHANCED_PREVIEW]: async () => {},
  },
  preSetupHook: async (opts) => {
    await ENGINE_HOOKS.setupBasic({ ...opts, extra: { idv2: true } });
    await NoteTestUtilsV4.modifyNoteByPath(
      { wsRoot: opts.wsRoot, vault: opts.vaults[0], fname: "foo" },
      (note: NotePropsV2) => {
        note.body = `[[bar]]`;
        return note;
      }
    );
  },
});

const NOTE_REF_RECURSIVE_BASIC_WITH_REHYPE = createProcTests({
  name: "NOTE_REF_RECURSIVE_WITH_REHYPE",
  setupFunc: async ({ engine, vaults, extra }) => {
    const proc = await MDUtilsV4.procFull({
      engine,
      wikiLinksOpts: { useId: true },
      dest: extra.dest,
      vault: vaults[0],
    });
    const txt = `((ref: [[foo.md]]))`;
    if (extra.dest === DendronASTDest.HTML) {
      const procRehype = MDUtilsV4.procRehype({ proc });
      const resp = await procRehype.process(txt);
      const respParse = await procRehype.parse(txt);
      const respTransform = await procRehype.run(respParse);
      return { resp, proc, respParse, respTransform };
    } else {
      const resp = await proc.process(txt);
      return { resp, proc };
    }
  },
  verifyFuncDict: {
    [DendronASTDest.MD_REGULAR]: async () => {},
    [DendronASTDest.HTML]: async ({ extra }) => {
      const { resp, respParse, respTransform } = extra;
      expect(resp).toMatchSnapshot();
      expect(respParse).toMatchSnapshot();
      expect(respTransform).toMatchSnapshot();
      expect(
        await AssertUtils.assertInString({
          body: resp.contents,
          match: [
            // link by id
            `<a href=\"foo-id.html\"`,
            // html quoted
            `Foo.One</h1>`,
            `Foo.Two</h1>`,
          ],
        })
      ).toBeTruthy();
    },
    [DendronASTDest.MD_ENHANCED_PREVIEW]: async () => {},
  },
  preSetupHook: async (opts) => {
    await ENGINE_HOOKS.setupNoteRefRecursive({
      ...opts,
      extra: { idv2: true },
    });
  },
});

const ALL_TEST_CASES = [
  ...WITH_VARIABLE,
  ...WITH_ASSET_PREFIX,
  ...WITH_ASSET_PREFIX_UNDEFINED,
  ...NOTE_REF_BASIC_WITH_REHYPE,
  ...NOTE_REF_RECURSIVE_BASIC_WITH_REHYPE,
  ...WITH_TITLE,
];

describe("MDUtils.proc", () => {
  test.each(
    ALL_TEST_CASES.map((ent) => [`${ent.dest}: ${ent.name}`, ent.testCase])
  )("%p", async (_key, testCase: TestPresetEntryV4) => {
    await runEngineTestV4(testCase.testFunc, {
      expect,
      createEngine,
      preSetupHook: testCase.preSetupHook,
    });
  });
});
