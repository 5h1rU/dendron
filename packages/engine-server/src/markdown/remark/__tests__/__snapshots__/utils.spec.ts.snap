// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MDUtils.proc "HTML: NOTE_REF_WITH_REHYPE": respRehype 1`] = `
VFile {
  "contents": "<p></p><p></p><div class=\\"portal-container\\">
<div class=\\"portal-head\\">
<div class=\\"portal-backlink\\">
<div class=\\"portal-title\\">From <span class=\\"portal-text-title\\">foo</span></div>
<a href=\\"foo-id.html\\" class=\\"portal-arrow\\">Go to text <span class=\\"right-arrow\\">â†’</span></a>
</div>
</div>
<div id=\\"portal-parent-anchor\\" class=\\"portal-parent\\" markdown=\\"1\\">
<div class=\\"portal-parent-fader-top\\"></div>
<div class=\\"portal-parent-fader-bottom\\"></div>
<p><a href=\\"bar.html\\">bar</a></p>
</div>    
</div><p></p><p></p>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`MDUtils.proc "HTML: WITH_FOOTNOTES" 1`] = `
VFile {
  "contents": "---
id: foo
title: Foo
desc: ''
updated: '1'
created: '1'
---

Here is a footnote reference,[^1]

[^1]: Here is the footnote.

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`MDUtils.proc "HTML: asset_prefix undefined" 1`] = `
VFile {
  "contents": "![alt-text](image-url.jpg)

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`MDUtils.proc "HTML: asset_prefix" 1`] = `
VFile {
  "contents": "![alt-text](/bond/image-url.jpg)

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`MDUtils.proc "MD_DENDRON: WITH_FOOTNOTES" 1`] = `
VFile {
  "contents": "---
id: foo
title: Foo
desc: ''
updated: '1'
created: '1'
---

Here is a footnote reference,[^1]

[^1]: Here is the footnote.

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`MDUtils.proc "MD_ENHANCED_PREVIEW: WITH_FOOTNOTES" 1`] = `
VFile {
  "contents": "---
id: foo
title: Foo
desc: ''
updated: '1'
created: '1'
---

Here is a footnote reference,[^1]

[^1]: Here is the footnote.

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`MDUtils.proc "MD_ENHANCED_PREVIEW: asset_prefix undefined" 1`] = `
VFile {
  "contents": "![alt-text](image-url.jpg)

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`MDUtils.proc "MD_ENHANCED_PREVIEW: asset_prefix" 1`] = `
VFile {
  "contents": "![alt-text](image-url.jpg)

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`MDUtils.proc "MD_REGULAR: WITH_FOOTNOTES" 1`] = `
VFile {
  "contents": "---
id: foo
title: Foo
desc: ''
updated: '1'
created: '1'
---

Here is a footnote reference,[^1]

[^1]: Here is the footnote.

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`MDUtils.proc "MD_REGULAR: asset_prefix undefined" 1`] = `
VFile {
  "contents": "![alt-text](image-url.jpg)

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`MDUtils.proc "MD_REGULAR: asset_prefix" 1`] = `
VFile {
  "contents": "![alt-text](image-url.jpg)

",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;
